docs/adr/0010-incoming-runtime-messages.md

```diff
@@ -0,0 +1,329 @@
+`executor` field of the runtime descriptor) as follows:
+
+```golang
+type ExecutorParameters struct {
+    // ... existing fields omitted ...
+
+    // MaxInMessages specifies the maximum size of the incoming message queue
+    // for this runtime.
+    MaxInMessages uint32 `json:"max_in_messages,omitempty"`
+}
+```
+
+### State
+
+This proposal introduces/updates the following consensus state in the roothash
+module:
+
+- **Incoming message queue metadata (`0x27`)**
+
+  Metadata for the incoming message queue.
+
+  ```
+  0x27 <H(runtime-id) (hash.Hash)>
+  ```
+
+  The value is the following CBOR-serialized structure:
+
+  ```golang
+  type IncomingMessageQueue struct {
+      // Size contains the current size of the queue.
+      Size uint32 `json:"size,omitempty"`
+
+      // NextSequenceNumber contains the sequence number that should be used for
+      // the next queued message.
+      NextSequenceNumber uint64 `json:"next_sequence_number,omitempty"`
+  }
+  ```
+
+- **Incoming message queue item (`0x28`)**
+
+  A queue of incoming messages pending to be delivered to the runtime in the
+  next round.
+
+  ```
+  0x28 <H(runtime-id) (hash.Hash)> <sequence-no (uint64)>
+  ```
+
+  The value is a CBOR-serialized `IncomingMessage` structure.
+
+### Transaction Methods
+
+This proposal updates the following transaction methods in the roothash module:
+
+#### Submit Message
+
+The submit message method allows anyone to submit incoming runtime messages to
+be queued for delivery to the given runtime.
+
+**Method name:**
+
+```
+roothash.SubmitMsg
+```
+
+**Body:**
+
+```golang
+type SubmitMsg struct {
+    ID     common.Namespace  `json:"id"`
+    Tokens quantity.Quantity `json:"tokens,omitempty"`
+    Data   []byte            `json:"data,omitempty"`
+}
+```
+
+**Fields:**
+
+- `id` specifies the destination runtime's identifier.
+- `tokens` specifies any tokens to be sent into the runtime as part of the
+  message being sent. The tokens are transferred before the message is processed
+  by the runtime.
+- `data` arbitrary data to be sent to the runtime for processing.
+
+The transaction signer implicitly specifies the caller. Upon executing the
+submit message method the following actions are performed:
+
+- Gas is accounted for (new `submitmsg` gas operation).
+
+- The runtime descriptor for runtime `id` is retrieved. If the runtime does not
+  exist or is currently suspended the method fails with `ErrInvalidRuntime`.
+
+- The `executor.max_in_messages` field in the runtime descriptor is checked. If
+  it is equal to zero the method fails with `ErrIncomingMessageQueueFull`.
+
+- The incoming queue metadata structure is fetched. If it doesn't yet exist it
+  is populated with zero values.
+
+- If the value of the `size` field in the metadata structure is equal to or
+  larger than the value of the `executor.max_in_messages` field in the runtime
+  descriptor the method fails with `ErrIncomingMessageQueueFull`.
```

Comment:
what is the gas cost of queuing a message?  the `max_in_messages` limit excessive queuing, but can be a DOS opportunity.  what can/should we do to prevent one runtime DOS'ing another?  enforce a min gas cost but allow recipient to increase? increase cost as queue gets longer?

Reply:
Currently the proposal has a fixed gas cost (see a few steps above) for queuing a message. But hm maybe it would make sense to dynamically increase the cost based on queue size.